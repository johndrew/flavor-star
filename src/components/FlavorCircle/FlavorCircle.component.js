import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Circle from '../Circle/Circle.component';

export default class FlavorCircle extends Component {

    render() {

        const {
            circle_color,
            label_color,
            label_secondary_text,
            label_text,
        } = this.getCircleDetails(this.props.type);
        const labelX = this.props.x / this.props.radius;
        const labelY = this.props.y / this.props.radius + 20;
        const labelSize = this.props.radius / 3 + 'px';
        return (
            <g>
                <Circle
                    color={circle_color}
                    grow_duration="500"
                    height={this.props.radius * 2}
                    label_details={{
                        text: label_text,
                        text_color: label_color,
                        text_size: labelSize,
                        secondary_text: label_secondary_text,
                        x: labelX,
                        y: labelY,
                    }}
                    radius={this.props.radius}
                    width={this.props.radius * 2}
                    x={this.props.x}
                    y={this.props.y} />
                }
            </g>
        )
    }

    /**
     * Retrieves details for context and style of a circle depending on the flavor type
     * @typedef {object} LabelDetail
     * @prop {string} circle_color The background color of the circle
     * @prop {string} label_color The color to display the label with. This should complement the
     * @prop {string} [label_secondary_text] A second label to display under the main label. This should be part of the main label semantically. This is not subtext
     * @prop {string} label_text
     * @param {string} type A supported flavor type
     * @returns {LabelDetail}
     * @throws
     */
    getCircleDetails(type) {

        switch (type) {
            case 'spice': return {
                circle_color: '#f05332',
                label_color: '#fef3ef',
                label_text: 'Spice',
            };
            case 'salty_umami': return {
                circle_color: '#ffdf01',
                label_color: '#786c32',
                label_secondary_text: 'Umami',
                label_text: 'Salty/',
            };
            case 'sweet': return {
                circle_color: '#7f3a66',
                label_color: '#fef3ef',
                label_text: 'Sweet',
            };
            case 'sour': return {
                circle_color: '#6abe45',
                label_color: '#fef3ef',
                label_text: 'Sour',
            };
            case 'bitter': return {
                circle_color: '#5c828f',
                label_color: '#fef3ef',
                label_text: 'Bitter',
            };
            default: throw new Error(`Type ${type} is not supported`);
        }
    }
}

FlavorCircle.propTypes = {
    type: PropTypes.string,
    radius: PropTypes.number,
    x: PropTypes.number,
    y: PropTypes.number,
};