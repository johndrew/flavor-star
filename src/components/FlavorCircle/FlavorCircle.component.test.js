import React from 'react';
import { shallow, mount } from 'enzyme';
import FlavorCircle from './FlavorCircle.component';

describe('<FlavorCircle />', () => {

    let requiredProps;
    beforeEach(() => {
        
        requiredProps = {
            type: 'spice', // no default color
        };
    });
    
    it('should render', () => {
        
        expect(shallow(<FlavorCircle {...requiredProps} />)).toMatchSnapshot();
    });

    describe('when setting the color prop', () => {

        describe('Positive Tests', () => {
        
            xit('should render red', () => {
                
                const wrapper = mount(<FlavorCircle
                    type="spice"
                    radius={10} />);
                expect(wrapper.find('svg').prop('fill')).toEqual(0);
            });

            xit('should render purple', () => {
                
                shallow(<FlavorCircle type="sweet" />);
            });
        });

        describe('Negative Test', () => {
            
            it('should error if color is not supported', () => {
                
                expect(() => shallow(<FlavorCircle type="pink" />)).toThrowError();
            });
        });
    });
});