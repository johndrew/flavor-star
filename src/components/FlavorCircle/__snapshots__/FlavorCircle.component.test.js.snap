// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlavorCircle /> should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FlavorCircle
    type="spice"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Circle
          color="#f05332"
          grow_duration="500"
          height={NaN}
          label_details={
            Object {
              "secondary_text": undefined,
              "text": "Spice",
              "text_color": "#fef3ef",
              "text_size": "NaNpx",
              "x": NaN,
              "y": NaN,
            }
          }
          width={NaN}
        />,
        "}",
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "#f05332",
          "grow_duration": "500",
          "height": NaN,
          "label_details": Object {
            "secondary_text": undefined,
            "text": "Spice",
            "text_color": "#fef3ef",
            "text_size": "NaNpx",
            "x": NaN,
            "y": NaN,
          },
          "radius": undefined,
          "width": NaN,
          "x": undefined,
          "y": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "}",
    ],
    "type": "g",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Circle
            color="#f05332"
            grow_duration="500"
            height={NaN}
            label_details={
              Object {
                "secondary_text": undefined,
                "text": "Spice",
                "text_color": "#fef3ef",
                "text_size": "NaNpx",
                "x": NaN,
                "y": NaN,
              }
            }
            width={NaN}
          />,
          "}",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": "#f05332",
            "grow_duration": "500",
            "height": NaN,
            "label_details": Object {
              "secondary_text": undefined,
              "text": "Spice",
              "text_color": "#fef3ef",
              "text_size": "NaNpx",
              "x": NaN,
              "y": NaN,
            },
            "radius": undefined,
            "width": NaN,
            "x": undefined,
            "y": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "}",
      ],
      "type": "g",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
